src/Compilers/Core/Portable/Compilation/TypeDeclarationInfo.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis
+{
+    public readonly struct TypeDeclarationInfo
+    {
+        public TypeDeclarationInfo(string typeName, string namespaceName, TypeKind kind, Accessibility accessibility, int arity)
+        {
+            Name = typeName;
+            ContainingNamespace = namespaceName;
+            Kind = kind;
+            Accessibility = accessibility;
+            Arity = arity;
+        }
```

Comment:
this feels like it will cause a lot of allcoations each time.  it's akin to forcing all the symbols to be created for each type in the world (though this at least benefits from being a struct).  It feels lke this could be pointed to *by* a declaration, and thus woudl never need to be recreated unless a new declaration was made.  Furthermore, it could just *point* at a Declaration instance, and thus acquire this data directly from that instance.

it would basically be a way of exposing a minimal surface area over a declaration node.

Reply:
ok , I think there's a few alternatives we can try
1. As you suggested here, let the new `TypeDeclarationInfo` points to the declaraion (but I think we still need `ContainingNamespace` property, since we can't get to its parent)
2. Bypass the object creation completely. With a method like below, we can create a completion item directly (where `ITypeDeclaration` is a new interface implemented by type declaration in declaration table, so no allocation there)

```cs
Task<ImmutableArray<T>>  VisitTopLevelTypeDelarations(Func<bool, ITypeDeclaration> predicate,  Func<T, ITypeDeclaration> create)
```